File: packages/settings/ios/Sources/SettingsPlugin/SettingsError.swift
import Foundation

/**
 Native error model for the Settings plugin (iOS).

 This enum represents all possible error categories
 produced by the native implementation.

 IMPORTANT:
 - Must NOT reference Capacitor
 - Must NOT reference JavaScript
 - Mapping to JS-facing error codes happens ONLY in the Plugin layer
 */
enum SettingsError: Error {
    case unavailable(String)
    case permissionDenied(String)
    case initFailed(String)
    case unknownType(String)

    /// Human-readable error message
    var message: String {
        switch self {
        case .unavailable(let message),
             .permissionDenied(let message),
             .initFailed(let message),
             .unknownType(let message):
            return message
        }
    }
}

File: packages/settings/android/src/main/java/io/capkit/settings/SettingsError.kt
package io.capkit.settings

/**
 * Native error model for the Settings plugin (Android).
 *
 * IMPORTANT:
 * - Extends Throwable to allow usage with Result.failure(...)
 * - Must NOT reference Capacitor or JavaScript
 * - Mapping to JS error codes happens ONLY in the Plugin layer
 */
sealed class SettingsError(
  message: String,
) : Throwable(message) {
  /** Feature or capability is not available on this device */
  class Unavailable(message: String) : SettingsError(message)

  /** Permission was denied by the user or system */
  class PermissionDenied(message: String) : SettingsError(message)

  /** Plugin failed to initialize correctly */
  class InitFailed(message: String) : SettingsError(message)

  /** Unsupported or unknown type/value was provided */
  class UnknownType(message: String) : SettingsError(message)
}
